### ISSUE TITLE
crop error on datageneration deep learning

### ISSUE THREAD

**Issue by vankhoa21991** [libvips/pyvips#166]
pyvips.error.Error: Caught Error in DataLoader worker process 3.
Original Traceback (most recent call last):
  File "/home/vankhoa/anaconda3/envs/py36con/lib/python3.6/site-packages/torch/utils/data/_utils/worker.py", line 178, in _worker_loop
    data = fetcher.fetch(index)
  File "/home/vankhoa/anaconda3/envs/py36con/lib/python3.6/site-packages/torch/utils/data/_utils/fetch.py", line 44, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/vankhoa/anaconda3/envs/py36con/lib/python3.6/site-packages/torch/utils/data/_utils/fetch.py", line 44, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/vankhoa/code/Bayer/MILPath/utils/datagen.py", line 57, in __getitem__
    img = utils.read_region(self.slides[slideIDX], coord, self.level, self.size)
  File "/home/vankhoa/code/Bayer/MILPath/utils/utils.py", line 49, in read_region
    img_vips = slide.copy().crop(coord[0], coord[1], size, size)
  File "/home/vankhoa/anaconda3/envs/py36con/lib/python3.6/site-packages/pyvips/vimage.py", line 805, in call_function
    return pyvips.Operation.call(name, self, *args, **kwargs)
  File "/home/vankhoa/anaconda3/envs/py36con/lib/python3.6/site-packages/pyvips/voperation.py", line 215, in call
    raise Error('unable to call {0}'.format(operation_name))
pyvips.error.Error: unable to call crop
  extract_area: bad extract area

Hello, this is the error I received when load tif pathology slide, then crop it to get the tiles used to train batch for deep learning. I verified that the coord are within the image. The similar approach using openslide.read_region works well. Could you give me your idea about this? Thanks. 
  

Process finished with exit code 1

**Comment by jcupitt** [comment]
Sorry, you need to give more information. What platform, sample code, sample image, what libvips version, how you built libvips etc. etc.

**Comment by vankhoa21991** [comment]
It's on ubuntu 18.04, anaconda3, libvips 8.8.3, I think I used anaconda packages. Image is a tif file, creates by pyvips.read_from_file(.svs) and then write_to_file tif.

**Comment by jcupitt** [comment]
`crop` is too slow for training. I would `Region.fetch` directly from the SVS file. 

```python
image = pyvips.Image.new_from_file('something.svs')
region = pyvips.Region.new(image)
bytes = region.fetch(0, 0, 100, 100)
```

It should be very quick.

**Comment by vankhoa21991** [comment]
How can I convert the region bytes to pil image?

**Comment by jcupitt** [comment]
Why do you need to do that? Could you explain what you are implementing?

**Comment by vankhoa21991** [comment]
It's a preprocessing stage for deep learning model, so the slides are read, and then tiles are extracted, each tile is the input to the model. So the slide is import with

slide = pyvips.Image.openslideload(slide_path, level=0)

Then the tiles are read with 
def read_region(slide, coord, level, size):
        img_vips = slide.copy().crop(coord[0], coord[1], size, size)
        np_3d = vips2numpy(img_vips)
        img = Image.fromarray(np_3d).convert("RGB")

**Comment by jcupitt** [comment]
But why do you need a PIL image? Are you writing to a file, or is that what your ML system needs? Does your ML system have any other way to pass pixels in?

**Comment by vankhoa21991** [comment]
Because later the image will have other preprocess steps, so a numpy array or PIL image are expected

**Comment by jcupitt** [comment]
Ah OK. I would use:

```python
data = region.fetch(0, 0, 100, 100)
np_array = np.ndarray(buffer=data, dtype=np.uint8, shape=[100, 100, 4])
```

I suppose you are doing flips etc. before passing the arrays to your ML system, is that right? You can probably get a speedup if you do the flips in pyvips.

**Comment by vankhoa21991** [comment]
This is great and very fast, for now in small scale it works fine. I will give you a feedback if it has troble in large scale. Thanks a lot.

**Comment by jcupitt** [comment]
You should be able to get more speed by doing the flips in pyvips, if that's what you are doing. Demo code:

https://github.com/libvips/pyvips/issues/100#issuecomment-493960943

**Comment by vankhoa21991** [comment]
pyvips.error.Error: Caught Error in DataLoader worker process 1.
Original Traceback (most recent call last):
  File "/home/vankhoa/anaconda3/envs/py36con/lib/python3.6/site-packages/torch/utils/data/_utils/worker.py", line 178, in _worker_loop
    data = fetcher.fetch(index)
  File "/home/vankhoa/anaconda3/envs/py36con/lib/python3.6/site-packages/torch/utils/data/_utils/fetch.py", line 44, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/vankhoa/anaconda3/envs/py36con/lib/python3.6/site-packages/torch/utils/data/_utils/fetch.py", line 44, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/vankhoa/code/Bayer/MILPath/utils/datagen.py", line 57, in __getitem__
    img = utils.read_region(self.slides[slideIDX], coord, self.level, self.size)
  File "/home/vankhoa/code/Bayer/MILPath/utils/utils.py", line 54, in read_region
    data = region.fetch(coord[0], coord[1], size, size)
  File "/home/vankhoa/anaconda3/envs/py36con/lib/python3.6/site-packages/pyvips/vregion.py", line 71, in fetch
    raise Error('unable to fetch from region')
pyvips.error.Error: unable to fetch from region
  
Process finished with exit code 1

Hello, I still face this issue when running in parallel. Reread the same tile only doesn't give any error.

**Comment by jcupitt** [comment]
No, it won't work in parallel, openslide is single-threaded. 

You could put a tilecache after openslideload. That will only let one thread at a time through to openslide.

**Comment by KVishnuVardhanR** [comment]
Hi,
even I got the same error, when I'm working with whole slide images,

`import pyvips
image = pyvips.Image.new_from_file('63.tif', access='sequential')
region = pyvips.Region.new(image)

x1,y1,x2,y2 =  26519, 11709, 46406, 27361 # coordinates     # 34007 60797 width, height

data = region.fetch(x1,y1,x2,y2)
cropped_image = np.ndarray(buffer=data, dtype=np.uint8)`

----> 8 data = region.fetch(x1,y1,x2,y2)
      9 cropped_image = np.ndarray(buffer=data, dtype=np.uint8)

/opt/conda/lib/python3.7/site-packages/pyvips/vregion.py in fetch(self, x, y, w, h)
     69         pointer = vips_lib.vips_region_fetch(self.pointer, x, y, w, h, psize)
     70         if pointer == ffi.NULL:
---> 71             raise Error('unable to fetch from region')
     72 
     73         pointer = ffi.gc(pointer, glib_lib.g_free)

Error: unable to fetch from region

**Comment by jcupitt** [comment]
`fetch` takes left, top, width, height.

**Comment by KVishnuVardhanR** [comment]
so, can you please give me the corrected code, that can be passes through fetch!!

**Comment by KVishnuVardhanR** [comment]
I tried the following code
```
def convert(size, box):
    dw = 1./size[0]
    dh = 1./size[1]
    x = (box[0] + box[1])/2.0
    y = (box[2] + box[3])/2.0
    w = box[1] - box[0]
    h = box[3] - box[2]
    x = x*dw
    w = w*dw
    y = y*dh
    h = h*dh
    return (x,y,w,h)

x,y,w,h = convert((34007, 60797),(26519, 11709, 46406, 27361))
data = region.fetch(int(x),int(y),int(w),int(h))
```
but the code gave this error:
Error                                     Traceback (most recent call last)
/tmp/ipykernel_17/1757444077.py in <module>
     20 
     21 x,y,w,h = convert((image.width, image.height),(26519, 11709, 46406, 27361))
---> 22 data = region.fetch(int(x),int(y),int(w),int(h))
     23 cropped_image = np.ndarray(buffer=data, dtype=np.uint8)

/opt/conda/lib/python3.7/site-packages/pyvips/vregion.py in fetch(self, x, y, w, h)
     69         pointer = vips_lib.vips_region_fetch(self.pointer, x, y, w, h, psize)
     70         if pointer == ffi.NULL:
---> 71             raise Error('unable to fetch from region')
     72 
     73         pointer = ffi.gc(pointer, glib_lib.g_free)

Error: unable to fetch from region
  VipsRegion: valid clipped to nothing

**Comment by jcupitt** [comment]
> 26519, 11709, 46406, 27361 # coordinates     # 34007 60797 width, height

Is this correct? A numpy array that's 46,406 x 27,361 pixels will be over 5GB. This will cause terrible performance problems.

`fetch` is designed for small patches, probably less than 128 x 128 pixels.

Could you explain what you are trying to achieve? What kind of model are you training? What tiles are you making?

**Comment by jcupitt** [comment]
Sample read of a large area:

```
$ python3
Python 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvips
>>> x = pyvips.Image.new_from_file("CMU-1.svs")
>>> x.width, x.height
(46000, 32914)
>>> reg = pyvips.Region.new(x)
>>> data = reg.fetch(100, 100, 44000, 31000)
>>> len(data)
5456000000
>>> 
>>> import numpy as np
>>> np_array = np.ndarray(buffer=data, dtype=np.uint8, shape=[44000, 31000, 4])
>>> 
>>> 
>>> np_array
array([[[243, 243, 243, 255],
        [243, 243, 243, 255],
        [243, 243, 243, 255],
        ...,
        [246, 243, 243, 255],
        [246, 243, 243, 255],
        [246, 243, 243, 255]]], dtype=uint8)
```

So it works, but performance is just horrible. `crop` and `write_to_memory` would work much better.

**Comment by KVishnuVardhanR** [comment]
I'm working on a cell classification problem using whole slide images and Im trying to crop the patches of the image which contains the cell, although only constraint is that I have utilize no more than 12GB RAM.  so far I'm able to read images of size (40k,50k), and able to find the coordinates of the region that I want to crop

so, the region corrdinates that I extracted are of form x1,y1,x2,y2.
generally in opencv to crop the ROI, we use `image[x1:x2, y1:y2]`, but I want to do the same using pyvips

**Comment by KVishnuVardhanR** [comment]
Can you please look at the coordinates and image width and height, and tell me how to extract the ROI using pyvips!!

x1,y1,x2,y2 = 26519, 11709, 46406, 27361
image width, image height = 34007 60797

**Comment by jcupitt** [comment]
It looks like your image is 340007 by 60797 pixels, but you are trying to extract a much larger area. But you've given so little information it's hard to be sure. And you shouldn't be using `fetch` anyway, or probably not. I can't really tell there either because you've also shared almost no details about what you're doing. How can I help if you don't explain the problem?

**Comment by jcupitt** [comment]
Please open a new issue, include full runnable code and a sample image, explain what you are trying to achieve, and give details about the platform and the library versions.
